hamp_radar/writezarr.py

```diff
@@ -0,0 +1,83 @@
+import xarray as xr
+from pathlib import Path
+from typing import Optional
+
+
+def ideal_zarrchunks(ds: xr.Dataset):
+    """returns ideal size of chunks for variables with these dimensions
+    so that chuunks are about 1MB. Addiontally cocx dimension (size=2)
+    is split into 2 seperate chunks and range dimension (size=512)
+    is sub-divided into 16 smaller chunks (size=32)"""
+    import warnings
+
+    frame = ds.sizes["frame"]
+    if frame > 2621440:
+        warnings.warn(
+            "Warning: number of frames is very large, consider changing ideal chunkshapes",
+            UserWarning,
+        )
+    return {
+        (): (),
+        ("frame",): (frame,),
+        ("frame", "cocx"): (frame, 1),
+        ("frame", "range", "cocx"): (8192, 32, 1),
+        ("frame", "range", "cocx", "fft", "iq"): (512, 32, 1, 256, 2),
+    }
+
+
+def zarr_chunks_encoding(ds: xr.Dataset):
+    ideal_chunks = ideal_zarrchunks(ds)
+    encoding = {
+        x: {"chunks": ideal_chunks[ds[x].dims]}
+        for x in list(ds.coords) + list(ds.data_vars)
+        if ds[x].chunks
+    }
+    return encoding
+
+
+def merge_encodings(e1: dict, e2: dict):
+    for x in set(e1.keys()) & set(e2.keys()):
+        e1[x].update(e2[x])
+    for x in set(e2.keys()) - set(e1.keys()):
+        e1[x] = e2[x]
+    return e1
+
+
+def zarr_encoding(ds: xr.Dataset, encoding: Optional[dict] = None):
+    if encoding is None:
+        return zarr_chunks_encoding(ds)
+    else:
+        return merge_encodings(encoding, zarr_chunks_encoding(ds))
+
+
+def prepare_daskarrays_for_zarrchunks(ds: xr.Dataset, encoding: dict):
+    def safe_chunks(current_chunks, desired_chunks):
+        return [max(i, c) for i, c in zip(current_chunks, desired_chunks)]
```

Comment:
What exactly is the purpose of `safe_chunks`? It it suppost to find a good common chunking for the combination of `current_chunks` and `desired_chunks`? If so, you may want to search for the [least common multiple](https://en.wikipedia.org/wiki/Least_common_multiple).

Reply:
Oh I was aiming to make the new chunks at least as big as the `desired_chunks`, but now I see that's wrong and actually we want the least common multiple (I didn't consider the case when it wasn't assumed that `desired_chunks` was already a factor/multiple of the `current_chunks`). I'll redo this.